java-checks/src/main/java/org/sonar/java/checks/security/CipherBlockChainingCheck.java

```diff
@@ -0,0 +1,98 @@
+import java.util.Optional;
+import javax.annotation.CheckForNull;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.java.matcher.MethodMatcher;
+import org.sonar.plugins.java.api.semantic.Symbol;
+import org.sonar.plugins.java.api.tree.ExpressionStatementTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.IdentifierTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.NewClassTree;
+import org.sonar.plugins.java.api.tree.StatementTree;
+import org.sonar.plugins.java.api.tree.Tree;
+
+@Rule(key = "S3329")
+public class CipherBlockChainingCheck extends AbstractMethodDetection {
+
+  @Override
+  protected List<MethodMatcher> getMethodInvocationMatchers() {
+    return ImmutableList.of(
+      MethodMatcher.create().typeDefinition("javax.crypto.spec.IvParameterSpec").name("<init>").withAnyParameters());
+  }
+
+  @Override
+  protected void onConstructorFound(NewClassTree newClassTree) {
+    Tree mTree = findParentMethod(newClassTree);
+    if (mTree.is(Tree.Kind.METHOD)) {
+      Optional<StatementTree> secureRandomOptional = findSecureRandomInMethod((MethodTree) mTree, returnSymbolOfArgument(newClassTree.arguments().get(0)));
+      if (!secureRandomOptional.isPresent()) {
+        reportIssue(newClassTree, "Use a dynamically-generated, random IV.");
+      }
+    }
+  }
+
+  private static Optional<StatementTree> findSecureRandomInMethod(MethodTree methodTree,
+                                                                  @javax.annotation.Nullable Symbol argumentSymbol) {
+    if (methodTree.block() != null && argumentSymbol != null) {
+      return methodTree.block().body().stream().filter(statementTree -> {
+        if (statementTree.is(Tree.Kind.EXPRESSION_STATEMENT)) {
+          ExpressionTree expr = ((ExpressionStatementTree) statementTree).expression();
+          if (expr.is(Tree.Kind.METHOD_INVOCATION)) {
+            MethodInvocationTree mit = ((MethodInvocationTree) expr);
+            if ("nextBytes".equals(mit.symbol().name())
+              && argumentSymbol.equals(returnSymbolOfArgument(mit.arguments().get(0)))) {
+              return true;
+            }
+          }
+        } 
+        return false;
+      }).findFirst();
+    }
+    return Optional.empty();
+
+  }
+
+  @CheckForNull
+  private static Symbol returnSymbolOfArgument(ExpressionTree argument) {
+    Symbol symbolArgument = null;
+    if (argument.is(Tree.Kind.IDENTIFIER)) {
+      symbolArgument = ((IdentifierTree) argument).symbol();
+    } else if (argument.is(Tree.Kind.MEMBER_SELECT)) {
+      symbolArgument = ((MemberSelectExpressionTree) argument).identifier().symbol();
+    }
+    return symbolArgument;
+  }
+
+  private static Tree findParentMethod(Tree tree) {
```

Comment:
This method could instead return a `MethodTree` or `null`. And it could be named `findEnclosingMethod`.